mips/ebase.S

```diff
@@ -29,59 +26,59 @@ _ebase:
         .local  invalid
         .local  pt_access
 
-tlb_refill:
-        # We don't want to generate TLB misses during tlb_refill.
+# TLB miss handler must avoid generating TLB miss during PT search.
+# 
+# Notes for the reader:
+#  - tlbr instruction overwrites EntryHi, so we have restore it,
+#  - handler does not fit into 256B, so it is splitted into two parts.
 
-        # Special case: an access was made to the PT from outside tlb_refill.
-        # This needs to be checked and handled separately.
+tlb_refill:
+        # Consider special case when an access to page table region was from
+        # the kernel (namely pmap module). As UPD & KPD are not mapped within
+        # page table we have to deal with it separately :(
 
         mfc0    k0, C0_BADVADDR
-        srl     k1, k0, 22
-        addi    k1, k1, -(PT_BASE >> 22)
+        srl     k1, k0, PDE_SHIFT
+        addi    k1, k1, -(PT_BASE >> PDE_SHIFT)
```

Comment:
I don't think the use of `PDE_SHIFT` is appropriate here. Sure, it **happens to have** the same value as what we need here, but we're shifting by 22 because 22 == `log2(PT_SIZE)`, not because 22 == `PDE_SHIFT`, and `log2(PT_SIZE)` may very well be different from `PDE_SHIFT`, depending on PTE size, address space size and page size. I agree that a hard-wired constant out of nowhere doesn't look good, so defining something like `PT_SIZE_SHIFT` or `LOG2_PT_SIZE` in `include/mips/pmap.h` and using that here seems more appropriate.

Reply:
I agree – done!